compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2798,3 +2851,230 @@ impl TyCategory {
         }
     }
 }
+
+impl<'tcx> InferCtxt<'_, 'tcx> {
+    pub fn find_block_span(&self, block: &'tcx hir::Block<'tcx>) -> Span {
+        let block = block.peel_blocks();
+        if let Some(expr) = &block.expr {
+            expr.span
+        } else if let Some(stmt) = block.stmts.last() {
+            // possibly incorrect trailing `;` in the else arm
+            stmt.span
+        } else {
+            // empty block; point at its entirety
+            block.span
+        }
+    }
+
+    pub fn find_block_span_from_hir_id(&self, hir_id: hir::HirId) -> Span {
+        match self.tcx.hir().get(hir_id) {
+            hir::Node::Block(blk) => self.find_block_span(blk),
+            // The parser was in a weird state if either of these happen...
```

Comment:
but does it really happen? could it be unreachable?

Reply:
I adapted other code that tries to handle this case gracefully, and don't really feel like introducing an ICE. Why does this matter?

Reply:
You are right, it doesn't.